015-kafka-topic-encryption.md

```diff
@@ -0,0 +1,169 @@
+The Encryption Module is the top-level component which encapsulates encryption functionality and is embedded in a proxy.
+A central design goal is adaptability through modularity. Encrypter/decrypter(s), the policy metadata service and the key management service (KMS) are expressed as interfaces and pluggable. The Encryption Module instantiates these components during initialization, allowing a configurable combination of KMS, metadata and encryption implementations.
+
+#### Encrypter-Decrypter
+The Encrypter-Decrypter sub-component performs the actual encryption and decryption of messages, implementing a particular encryption algorithm and making use of the KMS service to obtain topic keys.
+
+#### Key Service
+The Key Service is used in the Encryption Module to handle interactions with a key management system and to perform optional primitives such as those related to envelope encryption. The interface is generic in order to accommodate various key management implementations.
+
+#### Policy Service
+Encrypter-decrypters require a means to lookup encryption requirements and parameters for a topic. As a corresponding standard does not exist, the core component will interact with a generic internal interface for retrieving policy on how a topic is, or not is, to be encrypted. Implementers can develop and configure their own implementation of the policy service, be it a remote microservice, a local pre-configured cache, a cluster registry, etc.
+
+For each topic to be encrypted, a policy exists detailing:
+
+- topic name
+- encryption algorithm(s), cipher suite, and encryption/decryption steps
+- key management server address
+- key identifier
+- optional initialization information
+
+In our reference implementation, policies are encoded in JSON. An example topic encryption policy,
+used in testing, is shown below:
+
+```
+[
+ {
+   "topic" : "juicy-topic",
+   "kms_url" : "https://kms.abc.com",
+   "kms_instance": "11111111-2222-3333-4444-55555555555",
+   "key_ref" : "e2a73a6c4bf9",
+   "credential" : "dkT4WSCnrlNE7"
+ },
+ {
+   "topic" : "creditcard-data",
+   "kms_url" : "https://kms.abc.com",
+   "kms_instance": "11111111-2222-3333-4444-55555555555",
+   "key_ref" : "9957cdf5f32a",
+   "credential" : "dkT4WSCnrlNE7"
+ }
+]
+```
+
+In this example, two topics will be encrypted with two different keys residing in the same KMS.
+
+### Message Handling
+Message handling is concerned with facilities for intercepting messages, deserializing and inspecting them, invoking encryption as accorded by policy, modifying messages, and finally serializing and forwarding messages onward to Kafka brokers and clients. To accomplish encryption and decryption of data payloads, only Kafka *Produce requests* and *Fetch responses* need to be examined and potentially modified. High quality open source libraries exist already for implementing message introspection and serialization/deseriaization, e.g.,
+https://github.com/Shopify/sarama.
+
+### Message Encryption
```

Comment:
Maybe a stupid question, but I do not have much experience with encryption other then TLS connections. Is there any impact on this from things like:
* Expiration of the encryption keys?
* Keys needing to be replaced because they were leaked / compromised?

Or is this not something we need to address here?

Reply:
Partially. The idea is to let implementations of the key service interface deal with such requirements. I.e. implement envelope encryption to be able to expire root keys or support multiple different keys per topic.

I think one thing we need to mention is, that re-encryption of topics is not supported by this proposal. Meaning, if a key is compromised, data encrypted with it is compromised too and can only be mitigated by copying the topic and using a different encryption key.

Reply:
Yeah, I think it would be great to mention this to make it more clear.